// aspose.barcode/lib/AsposeBarcode.js
'use strict'
const java = require('java');
const joint_ = require('./Joint');
const complexbarcode_ = require("./ComplexBarcode");
const generation_ = require("./Generation");
const recognition_ = require("./Recognition");

function pushJar()
{
    java.classpath.push(__dirname + "/aspose-barcode-nodejs-22.6.jar");
}
pushJar();

module.exports = { complexbarcode_, recognition_, generation_,joint_,
    BarcodeGenerator : generation_.BarcodeGenerator,
    BarcodeParameters: generation_.BarcodeParameters,
    BaseGenerationParameters : generation_.BaseGenerationParameters,
    BorderParameters : generation_.BorderParameters,
    ChecksumValidation:generation_.ChecksumValidation,
    CaptionParameters: generation_.CaptionParameters,
    Unit: generation_.Unit,
    Padding: generation_.Padding,
    CodetextParameters: generation_.CodetextParameters,
    PostalParameters: generation_.PostalParameters,
    AustralianPostParameters: generation_.AustralianPostParameters,
    CodablockParameters: generation_.CodablockParameters,
    DataBarParameters: generation_.DataBarParameters,
    DataMatrixParameters: generation_.DataMatrixParameters,
    PatchCodeParameters: generation_.PatchCodeParameters,
    Code16KParameters: generation_.Code16KParameters,
    DotCodeParameters: generation_.DotCodeParameters,
    ITFParameters: generation_.ITFParameters,
    QrParameters: generation_.QrParameters,
    Pdf417Parameters: generation_.Pdf417Parameters,
    SupplementParameters: generation_.SupplementParameters,
    MaxiCodeParameters: generation_.MaxiCodeParameters,
    AztecParameters: generation_.AztecParameters,
    CodabarParameters: generation_.CodabarParameters,
    CouponParameters: generation_.CouponParameters,
    FontUnit: generation_.FontUnit,
    ExtCodetextBuilder: generation_.ExtCodetextBuilder,
    QrExtCodetextBuilder: generation_.QrExtCodetextBuilder,
    QrStructuredAppendParameters: generation_.QrStructuredAppendParameters,
    BarcodeClassifications: generation_.BarcodeClassifications,
    FontStyle: generation_.FontStyle,
    CodabarSymbol: generation_.CodabarSymbol,
    DataMatrixEncodeMode: generation_.DataMatrixEncodeMode,
    BorderDashStyle: generation_.BorderDashStyle,
    ITF14BorderType: generation_.ITF14BorderType,
    QREncodeMode: generation_.QREncodeMode,
    DataMatrixEccType: generation_.DataMatrixEccType,
    QRVersion: generation_.QRVersion,
    AztecSymbolMode: generation_.AztecSymbolMode,
    Pdf417ErrorLevel: generation_.Pdf417ErrorLevel,
    Pdf417CompactionMode: generation_.Pdf417CompactionMode,
    QRErrorLevel: generation_.QRErrorLevel,
    QREncodeType: generation_.QREncodeType,
    CodabarChecksumMode: generation_.CodabarChecksumMode,
    CodeLocation: generation_.CodeLocation,
    FontMode: generation_.FontMode,
    TextAlignment: generation_.TextAlignment,
    AutoSizeMode: generation_.AutoSizeMode,
    GraphicsUnit: generation_.GraphicsUnit,
    EncodeTypes: generation_.EncodeTypes,
    MacroCharacter: generation_.MacroCharacter,
    PatchFormat: generation_.PatchFormat,
    ECIEncodings: generation_.ECIEncodings,
    EnableChecksum: generation_.EnableChecksum,
    BarCodeImageFormat: generation_.BarCodeImageFormat,
    BarCodeReader: recognition_.BarCodeReader,
    Quadrangle: recognition_.Quadrangle,
    QRExtendedParameters: recognition_.QRExtendedParameters,
    Pdf417ExtendedParameters: recognition_.Pdf417ExtendedParameters,
    OneDExtendedParameters: recognition_.OneDExtendedParameters,
    Code128ExtendedParameters: recognition_.Code128ExtendedParameters,
    BarcodeSvmDetectorSettings: recognition_.BarcodeSvmDetectorSettings,
    BarCodeResult: recognition_.BarCodeResult,
    BarCodeRegionParameters: recognition_.BarCodeRegionParameters,
    BarCodeExtendedParameters: recognition_.BarCodeExtendedParameters,
    QualitySettings: recognition_.QualitySettings,
    Code128DataPortion: recognition_.Code128DataPortion,
    DataBarExtendedParameters: recognition_.DataBarExtendedParameters,
    DecodeType: recognition_.DecodeType,
    Code128SubType: recognition_.Code128SubType,
    CustomerInformationInterpretingType: recognition_.CustomerInformationInterpretingType,
    BarCodeConfidence: recognition_.BarCodeConfidence,
    License:joint_.License,
    BarcodeException:joint_.BarcodeException,
    Rectangle:joint_.Rectangle,
    Point:joint_.Point,
    BuildVersionInfo:joint_.BuildVersionInfo,
    ComplexBarcodeGenerator : complexbarcode_.ComplexBarcodeGenerator,
    Address : complexbarcode_.Address,
    AddressType : complexbarcode_.AddressType,
    AlternativeScheme : complexbarcode_.AlternativeScheme,
    ComplexCodetextReader : complexbarcode_.ComplexCodetextReader,
    QrBillStandardVersion : complexbarcode_.QrBillStandardVersion,
    SwissQRBill : complexbarcode_.SwissQRBill,
    SwissQRCodetext : complexbarcode_.SwissQRCodetext}
